pkg/restic/commands.go

```diff
@@ -0,0 +1,178 @@
+	UID      int       `json:"uid"`
+	Gid      int       `json:"gid"`
+	Tags     []string  `json:"tags"`
+}
+
+func (w *ResticWrapper) listSnapshots(snapshotIDs []string) ([]Snapshot, error) {
+	result := make([]Snapshot, 0)
+	args := w.appendCacheDirFlag([]interface{}{"snapshots", "--json", "--quiet", "--no-lock"})
+	args = w.appendCaCertFlag(args)
+	for _, id := range snapshotIDs {
+		args = append(args, id)
+	}
+
+	err := w.sh.Command(Exe, args...).UnmarshalJSON(&result)
+	return result, err
+}
+
+func (w *ResticWrapper) deleteSnapshots(snapshotIDs []string) ([]byte, error) {
+	args := w.appendCacheDirFlag([]interface{}{"forget", "--quiet", "--prune"})
+	args = w.appendCaCertFlag(args)
+	for _, id := range snapshotIDs {
+		args = append(args, id)
+	}
+
+	return w.run(Exe, args)
+}
+
+func (w *ResticWrapper) initRepositoryIfAbsent() ([]byte, error) {
+	log.Infoln("Ensuring restic repository in the backend")
+	args := w.appendCacheDirFlag([]interface{}{"snapshots", "--json"})
+	args = w.appendCaCertFlag(args)
+	if _, err := w.run(Exe, args); err != nil {
+		args = w.appendCacheDirFlag([]interface{}{"init"})
+		args = w.appendCaCertFlag(args)
+
+		return w.run(Exe, args)
+	}
+	return nil, nil
+}
+
+func (w *ResticWrapper) backup(path string, tags []string) ([]byte, error) {
+	log.Infoln("Backing up target data")
+	args := []interface{}{"backup", path}
+	if w.config.Hostname != "" {
+		args = append(args, "--host")
+		args = append(args, w.config.Hostname)
+	}
+	// add tags if any
+	for _, tag := range tags {
+		args = append(args, "--tag")
+		args = append(args, tag)
+	}
+	args = w.appendCacheDirFlag(args)
+	args = w.appendCaCertFlag(args)
+
+	return w.run(Exe, args)
+}
+
+func (w *ResticWrapper) cleanup(cleanupOptions CleanupOptions) ([]byte, error) {
+	log.Infoln("Cleaning old snapshots according to retention policy")
+
+	args := []interface{}{"forget"}
+
+	policy := cleanupOptions.RetentionPolicyName
+	if !strings.HasPrefix(policy, "--") {
+		policy = "--" + policy
+	}
+
+	args = append(args, policy)
+	args = append(args, cleanupOptions.RetentionValue)
+
+	if cleanupOptions.Prune {
+		args = append(args, "--prune")
+	}
+
+	if cleanupOptions.DryRun {
+		args = append(args, "--dry-run")
+	}
+
+	if len(args) > 1 {
+		args = w.appendCacheDirFlag(args)
+		args = w.appendCaCertFlag(args)
+
+		return w.run(Exe, args)
+	}
+	return nil, nil
+}
+
+func (w *ResticWrapper) restore(path, host, snapshotID string) ([]byte, error) {
+	log.Infoln("Restoring backed up data")
+	args := []interface{}{"restore"}
+	if snapshotID != "" {
+		args = append(args, snapshotID)
+	} else {
+		args = append(args, "latest")
+	}
+	args = append(args, "--path")
+	args = append(args, path) // source-path specified in restic fileGroup
+	args = append(args, "--host")
+	args = append(args, host)
+
+	// Remove last part from the path.
+	// https://github.com/appscode/stash/issues/392
+	args = append(args, "--target")
+	args = append(args, filepath.Dir(path))
```

Comment:
Restic maintains source path while restoring backups. Shouldn't the target be just `/` ?